### ISSUE TITLE
Feature request: hyperlinks on text

### ISSUE THREAD

**Issue by RackhamLeNoir** [citation-style-language/schema#145]
Hi,

Hyperlinks are convenient in bibliographies to access the paper directly. However including the url as a text uses space, and is against what hyperlinks are all about. I would like to be able to add a hyperlink to the url field in the bibtex entry to the title for example. As far as I understood it is not currently possible with CSL.

Is it possible to add this feature in the future? Either add a url property to text, or add a sub-item url.

Thanks,

**Comment by rmzelle** [comment]
It's definitely worth considering. (a related feature request is the ability to automatically hyperlink DOI links and other URLs)

If you are aware of any official style guides that specify that specific section of a bibliographic entry need to be hyperlinked, please let us know.

**Comment by RackhamLeNoir** [comment]
I personnally use ACM SIGCHI templates, usually with latex. In our group we used to add hyperlinks to the url if present in the bibtex file, with a couple of modifications in the bst file, such as:

```
FUNCTION {format.title}
{ url empty$
    { format.title.orig }
    { new.block "\href{" url * "}{" * format.title.orig * "}" * }
  if$
}
```

```
FUNCTION {inproceedings}
{ output.bibitem
  format.authors "author" output.check
  new.block
  format.title "title" output.check
  new.block
  crossref missing$
    { format.in.proc.date "booktitle" output.check
      format.ineditors output
      format.bvolume output
      format.number.series output
      organization output
      publisher output
      format.addr.date
    }
    { format.incoll.inproc.crossref output.nonnull }
  if$
  format.pp.pages output
  new.block
  note output
  fin.entry
}
```

These are not official templates, but modifications of them. But I think official templates will take benefit of such features only if the citation description languages propose them.

**Comment by dhimmel** [comment]
I agree that hyperlinks in references are important, and fit well within the scope of what CSL should be able to do. However, it's not currently part of CSL as per discussion in https://github.com/Juris-M/citeproc-js-docs/issues/15:

> it's not possible to produce links with CSL style code alone. You need to hack the processor directly

**Comment by danmackinlay** [comment]
> It's definitely worth considering. (a related feature request is the ability to automatically hyperlink DOI links and other URLs)
> 
> If you are aware of any official style guides that specify that specific section of a bibliographic entry need to be hyperlinked, please let us know.

This is IMO bigger than journal style guides. (Not that I am suggesting you are arguing that it is purely that, @rmzelle :smile: :)

Now that CSL is the standard citation processor for projects like pandoc via [pandoc-citeproc](https://github.com/jgm/pandoc-citeproc) which is in turn the backend for popular academic  blogging package [blogdown](https://github.com/rstudio/blogdown). These tools, and others like them, are rendering citations for online consumption, not for printing out, and basic hyperlinking functionality would seem wise here. In the setting of blogs, this is not just for the sake of meeting the taste of journal publishers/editors, but also rather the standards of the internet generally where the lack of official written style guides does save us we are not still violating unwritten social norms and good usability by ignoring the existence of hyperlinks.

**Comment by bdarcus** [comment]
Just to give some context, I've always viewed as unclear whether this is
the proper scope for CSL, CSL implementors, or both. And in fact, I lean
towards the latter.

You could imagine preference settings on implementors that would cover the
important cases (doi and http hyperlinking).

Right?

And if that's the case, why add it to CSL?

**Comment by bdarcus** [comment]
Sorry, meant "lean towards" the implementors; not CSL.

**Comment by dhimmel** [comment]
> "lean towards" the implementors; not CSL

Leaving linking to the implementors does not allow for hyperlinking titles. It only allows hyperlinking identifiers that are explicitly inserted into the plain reference. Identifiers and URLs are often ugly. It's often nice to hide them in a hyperlink.

Also, CSL supports formatting like spacing, bold, italics. Hyperlinks are well within this feature scope.

**Comment by bdarcus** [comment]
As you imagine, can be a slippery slope.

I must have missed the requirement to be able to anchor links to arbitrary
variables (rather than just urls and dois).

Would help to have some documentation of the need (in style guides). Do we
have that?

**Comment by dhimmel** [comment]
> Would help to have some documentation of the need (in style guides).

We recently designed a modern style for Manubot manuscripts, where we would have liked to hyperlink the title but were limited by CSL. History at https://github.com/manubot/rootstock/issues/53.

**Comment by danmackinlay** [comment]
Wow @dhimmel Manubot looks neat.

> Would help to have some documentation of the need (in style guides). Do we have that?

Hmm. Tricky, @bdarcus. For HTML generally, so, but possibly not in the format of a journal. There _is_ a style guide for web citations, of sorts, which I guess might legitimate this use case, if style guides are what we wish to work from.

Specifically, the W3C recommends for semantic HTML the use of the  [cite element](https://www.w3.org/TR/html52/textlevel-semantics.html#the-cite-element). They do not require that URLs be linked rather than spelled out, but they do _allow_ it.

Here is an example of an unlinked HTML citation from that document:

```html
<p><cite>Universal Declaration of Human Rights</cite>, United Nations,
December 1948. Adopted by General Assembly resolution 217 A (III).</p>
```

Here is an example of a linked one:

```html
Comment by <cite><a href="https://oli.jp">Oli Studholme</a></cite>
  <time datetime="2013-08-19T16:01">August 19th, 2013 at 4:01 pm</time>
```

Although it is not a _semantic_ HTML requirement, the W3C's WCAG recommendation for accessible content [explicitly advises against](https://www.w3.org/TR/WCAG20-TECHS/G91.html) naked URLs as link text:

> The objective of this technique is to describe the purpose of a link in the text of the link. The description lets a user distinguish this link from links in the Web page that lead to other destinations and helps the user determine whether to follow the link. The URI of the destination is generally not sufficiently descriptive.

Note that there is an alternative way of meeting the accessibility guidelines, which is to determine [link destination by context](https://www.w3.org/TR/2008/REC-WCAG20-20081211/#navigation-mechanisms-refs), which could arguably be satisfied by a typically generated bibliography; However in the case that a bibliography item has, e.g. a DOI URL and some other URL, one woudl once again be in violation.

Moe generally, the standard practice that link text is _not_ the URL when you are on the computer is so pervasive and universal that it is hard to find it documented because it is assumed.

However we can find implicit assumptions that support this as standard practice; link text not being the link URL is implicit in the common practice of [print](https://www.smashingmagazine.com/2018/05/print-stylesheets-in-2018/#inserting-content) [stylesheets](https://makandracards.com/makandra/29149-bootstrap-how-to-avoid-printing-link-urls) which display the URLs *as well as the link text* but only when printing, not when on a screen. There is, AFAICT, no reverse CSS trick, to banish length URLs from link text.

And most of all, of course,
This very site, github, for example, refers to

* this repository as "[schema](https://github.com/citation-style-language/schema)", rather than "[https://github.com/citation-style-language/schema](https://github.com/citation-style-language/schema)", and
* to this issue as [Feature request: hyperlinks on text](https://github.com/citation-style-language/schema/issues/145) rather than "[https://github.com/citation-style-language/schema/issues/145](https://github.com/citation-style-language/schema/issues/145)". 

This convention is of course different on facebook posts or twitter, but nonetheless is suggestive that for *some* websites at least the implied style does not include link-text-as-URL and we are currently incapable of supporting that.

**Comment by danmackinlay** [comment]
With reference to that style guide, here is an example of the kind of pathology I mean; this is a citation rendered on my blog using `apa.csl` via `pandoc-citeproc`. It is hard to imagine the URL, which takes up most of the entry, would not be better as a link whose target is less obvious.

![Screenshot from 2019-09-16 10-35-08](https://user-images.githubusercontent.com/21740/64929804-13b2d500-d86e-11e9-9262-d7e1c1b7ac82.png)

**Comment by danmackinlay** [comment]
With pandoc-citeproc this problem arises also. See https://github.com/jgm/pandoc-citeproc/issues/155 for anidentification of a style-file constraint and https://github.com/jgm/pandoc-citeproc/issues/318 for a workaround based on LaTeX macros, which AFAICS does not help anyone outside of the context of LaTeX (e.g. people producing HTML documents are still stranded)

**Comment by adam3smith** [comment]
Since this has come up again, I'll chime in. Perhaps even more strongly than @bdarcus , I'd much rather have this handled by implementers. Style proliferation is a serious issue for CSL and adding a feature that could arguably be used to created duplicates of hundreds of existing styles doesn't seem good for the project.

It also means that users who want a fairly standard feature added to an existing style will need to code CSL or ask (already overstretched) volunteers for help -- also not good.

I also disagree that it's the case that implementers couldn't handle standard cases such as linking titles. E.g.  ZotPress, the WordPress add-on for Zotero, allows the specification of a `urlwrap` variable for a bibliography. You can set that to title and then titles will be hyperlinked. Something along those line just seems like a much more versatile solution to me.

**Comment by danmackinlay** [comment]
**tl;dr**: IMO punting linking to implementers could either exclude many plausible types of linking, or lead to a shadow CSL for implementers to describe the various types of linking. There is the risk of complexity either way but I argue that overall we have less complexity if we keep it *in* CSL.

---

Good point @adam3smith  about keeping CSL compact and about keeping the implementation burden away from end-users, and about keeping the number of styles as low as possible. 
It is not clear to me, however, that the proposal you mention, of keeping any specification of linking out CSL, would necessarily achieve either of those. I suspect our different perspectives on this are  partly based on some details of our imagined use cases and what we should expect users to implement.

For one thing, we want to keep CSL simple by avoiding a proliferation of options for a commonly used function that are specifed in an ad hoc way by various implementers. `urlwrap` in Zotpress is a case in point. It is not clear to me that links should always be from the title (or collection title) to the url. Why should it not be to the DOI endpoint? Might it not the article link url be sometimes best represented as the link text, e.g when we are preparing a document both for web and print and do not have control of the stylesheet? Might we not sometimes wish to have the link be a link *from* the DOI, as in e.g. [Pubmed references](https://www.ncbi.nlm.nih.gov/pubmed/30467447), or from some other arbitrary link? Specifying these various behaviors would be natural from the CSL layer. If we punt it to the implementers AFAICS we must either

1. require implementers to individually and separately decide on what the "true" linking behaviour is for their implementation, as with zotpress, which might not meet all style guides, or 
2. create a miniature language specification that describes how a CSL implementer should inject various elements and attributes into a citation (which element to wrap, which URL to wrap with etc) , mirroring closely the facilities that CSL already provides.

Both these seem likely to add complexity overall and to lead to fracturing between implementations rather than agreement.

The point about proliferating styles is ambiguous. We have many styles in the styles repository because, (?) in principle, that is how many different styles that the various journal require, and as such we hope them to proliferate only if the journal style requirements change If the Rwandan Journal Of Optometry suddenly goes online-only and requires links in the article titles where available, then we could update `rwandan_journal_of_optometry.csl` accordingly, and in fact would need to. If some other journal dislikes links but otherwise recommends the style of the Rwandan Journal Of Optometry and wants to fork from the old version, then we have more-or-less the same process that currently produces hundreds of very similar styles and are not worse off. (I mean, is it? I'm making assumptions here) On the other hand, if we are producing CSL styles for informal publication contexts like blogs etc, those have a much weaker argument for inclusion in styles repository. There is little reason to include both `dans_style_for_blogs.csl` and `dans_style_for_blogs_with_smallcaps_title.csl`, although there is of course difficulty with gatekeeping here. Nonetheless I think the case that we would need to duplicate many journal styles to include links just to satisfy the vanity of an opinionated blogger is not strong  - and if an opinionated blogger has the time and desire to maintain their own special CSL file that is distinct from a handful of standard linked styles and NOT in the main repository, that is presumably their own business and even desirable.

If we are truly opposed to including linking semantics here there is another option which is even more flexible that we could accomplish the desired effects. We could for example, include in the CSL spec the ability to let the CSL style tag parts of the output citation with HTML5 microformats (for author, date title etc) and class attributes etc so that end users could post-process them with some HTML processing tool or XSLT or such to reliably extract the URL and render it as different HTML. This seems to be more complex  and is once again re-implementing a lot of the same infrastructure that is already implied by CSL.

Anyway, I fear I'm prattling on. I guess this boils down to whether we think we know what standard type of link users will need, and if that is trivial or whether it would be simplest if it had access to the power of CSL.

**Comment by danmackinlay** [comment]
Since this is a highly-ranked search result for this problem, I'm going to link here to [forum](https://discourse.citationstyles.org/t/linking-the-doi-variable/1074/25) [discussions](https://discourse.citationstyles.org/t/promoting-links-in-html-output/958) on this topic, as per the suggestion over in the citeproc-js tracker.

NB, some of those are more than 10 years old now, so I am nervous about bumping them. Or hoping for a speedy resolution ...

**Comment by bdarcus** [comment]
I'm forgetting how we're splitting where to put discussions; will defer to @adam3smith and @rmzelle. 

But I will say here, as I'm thinking about this again, there is a straightforward midway position on this question, which is that the CSL spec can itemize the linking use cases, and either make recommendations for implementers, or define requirements.

I would think that would be a necessary first step before adding full-blown link config to CSL (which I'm still skeptical we want to do), and would probably cover most cases anyway.

So my suggestion: someone (or ideally, a group of interested people on some collaborative platform) draft a proposal for a section or subsection of the spec that addresses this, for consideration.

A simple version could start with something like [this](https://discourse.citationstyles.org/t/promoting-links-in-html-output/958/2)?

**Comment by bwiernik** [comment]
I think link can go into a recommended API document as described here: https://github.com/citation-style-language/documentation/issues/72

Can we move this issue over to the documentation repo?
